Gems/Atom/Feature/Common/Code/Source/Mesh/ModelReloader.cpp

```diff
@@ -0,0 +1,175 @@
+        {
+            onReloadedEventHandler.Connect(m_onModelReloaded);
+        }
+
+        void ModelReloader::OnAssetReloaded(AZ::Data::Asset<AZ::Data::AssetData> asset)
+        {
+            DependencyList& pendingDependencies = GetPendingDependencyList();
+
+            const Data::AssetId& reloadedAssetId = asset.GetId();
+
+            // Find the index of the asset that was reloaded
+            const auto matchesId = [reloadedAssetId](const Data::Asset<Data::AssetData>& asset){ return asset.GetId() == reloadedAssetId;};
+            const auto& iter = AZStd::find_if(AZStd::begin(pendingDependencies), AZStd::end(pendingDependencies), matchesId);
+            AZ_Assert(
+                iter != AZStd::end(pendingDependencies),
+                "ModelReloader - handling an AssetReloaded event for an asset that is not part of the dependency list.");
+            size_t currentIndex = AZStd::distance(AZStd::begin(pendingDependencies), iter);
+
+            // Keep a reference to the newly reloaded asset to prevent it from being immediately released
+            pendingDependencies[currentIndex] = asset;
+            Data::AssetBus::MultiHandler::BusDisconnect(reloadedAssetId);
+
+            // Clear the bit, now that it has been reloaded
+            m_pendingDependencyListStatus.reset(currentIndex);
+
+            if (m_pendingDependencyListStatus.none())
+            {
+                AdvanceToNextLevelOfHierarchy();
+            }
+        }
+
+        void ModelReloader::OnAssetReloadError(Data::Asset<Data::AssetData> asset)
+        {
+            // An error is actually okay/expected in some situations.
+            // For example, if the 2nd UV set was removed, and we tried to reload the second uv set, the reload would fail.
+            // We want to treat it as a success, and mark that dependency as 'up to date'
+            OnAssetReloaded(asset);
+        }
+
+        void ModelReloader::InsertMeshDependencyIfUnique(Data::Asset<Data::AssetData> asset)
+        {
+            if (AZStd::find(AZStd::begin(m_meshDependencies), AZStd::end(m_meshDependencies), asset) == AZStd::end(m_meshDependencies))
+            {
+                // Multiple meshes may reference the same buffer, so only add the dependency if it is unique
+                m_meshDependencies.push_back(asset);
+            }
+        }
+
+        void ModelReloader::ReloadDependenciesAndWait()
+        {
+            // Get the current list of dependencies depending on the current state
+            DependencyList& dependencies = GetPendingDependencyList();
+
+            if (!m_pendingDependencyListStatus.none())
+            {
+                AZ_Assert(
+                    m_pendingDependencyListStatus.none(),
+                    "ModelReloader attempting to add new dependencies while still waiting for other dependencies in the hierarchy to "
+                    "load.");
+            }
```

Comment:
What's the point of the `if`? the assert is already checking `m_pendingDependencyListStatus` in its condition. Unless I'm missing something the `if` could be removed.

Reply:
I had this so that I could put a breakpoint on this line (Visual Studio claims that .none() has side effects, so I cannot use a conditional breakpoint.

I can remove the if now that I'm done debugging.